### ISSUE TITLE
`deno bundle`: Ordering of bundled code is incorrect?

### ISSUE THREAD

**Issue by ebebbington** [denoland/deno#11286]
First off, canary (bumping swc_bundler to 0.45.0) fixed the infinite loop issue for me, which is great! (https://github.com/denoland/deno/issues/8224)

But now, when i bundle my code, the ordering seems off, causing calls to uninitialised variables (eg incorrect hoisting)

An error message I see is: `Uncaught ReferenceError: Cannot access 'idEncoder' before initialization`

Which points to this inside the bundled file:

```js
const idEncoder = pseudoRandomEncode(2n ** 20n, 387420489n);
function* pseudoRandomIdGenerator() {
    let i = 0n;
    while(true){
        yield idEncoder(++i).toString(36);
    }
}
```

Long story short, my code that is bundled is placed somewhat 3/4 through the file, so in a nutshell, this is how the bundled file looks:

```js
myFn() // by the power of hoisting, i declare `name` uninitialised
const name = "ed"
function myFn() {
  return name 
}
```

**Comment by kitsonk** [comment]
We need to try to get to a minimal reproduction so we can open a concise issue with swc.

**Comment by ebebbington** [comment]
> We need to try to get to a minimal reproduction so we can open a concise issue with swc.

sure, i'll try get one sorted

**Comment by ebebbington** [comment]
I'm guessing this is why: `Using slow, fallback logic for topological sorting`

**Comment by ebebbington** [comment]
same issue with `Deno.emit` and setting `bundle` to `module`

**Comment by ebebbington** [comment]
to clarify @kitsonk, whatever (internally) makes the bundle use the "slow, fallback logic", is the culprit, a simple 'example' is below, granted this works, *but* if the slow fallback logic was in place, it would not work:

```ts
// mod.ts
import { register } from "./register.ts"
register()
```
```ts
// register.ts
import { id } from "./id.ts"
export function register () { return id() }
```
```ts
// id.ts
const pseudoNumber = Math.random()
export const id = () => pseudoNumber
```
```shell
$ deno bundle mod.ts mod.js
```

**Comment by ebebbington** [comment]
Also maybe this might help, i managed to 'fix' it (really its just a hack) by doing this:

![image](https://user-images.githubusercontent.com/47337480/124590668-51be2600-de53-11eb-8dac-9bf29d7f276c.png)

**Comment by dsherret** [comment]
@ebebbington can this one be closed? If not, would you be able to provide a minimal reproduction along with actual and expected outputs? I got this output which seems correct:

```ts
const pseudoNumber = Math.random();
const id = ()=>pseudoNumber
;
function register() {
    return id();
}
register();
```

**Comment by arnauduanra** [comment]
I have a codebase written for NodeJS. I am trying to run this codebase with Deno. To achieve this, I am using a map file which looks like:  
```
... 
"stream": "https://deno.land/std/node/stream.ts",
"util": "https://deno.land/std/node/util.ts",
"http": "https://deno.land/std/node/http.ts",
"fs": "https://deno.land/std/node/fs.ts",
"crypto": "https://deno.land/std/node/crypto.ts"
...
```

When I use `deno run ...`  my project is running well. But when I am using `deno compile ...` or `deno bundle ...`, it produces an error when I try to run the file produced:  

```
Uncaught ReferenceError: Cannot access 'NodeTypeError' before initialization. 
```

If I look into the bundle produced, I can see that 'NodeTypeError' is declared after the usage (as mentioned by the error).   

So this error of "ordering of bundled code" looks like a big issue.

**Comment by dsherret** [comment]
@arnauduanra if you get a chance, would you be able to provide a reproducible contained example? Otherwise it's a bit of a wild goose chase.

**Comment by audrow** [comment]
@dsherret, I believe that I'm having the same problem, I get ordering errors when bundling a small project. Here is the project: https://github.com/audrow/update-ros-pkg-versions.

I use Github Actions to bundle my project and put it in a build folder. You can see the bundled file here: https://raw.githubusercontent.com/audrow/update-ros-pkg-versions/deploy/build/update-ros-pkg-versions.js.

You can try to run the bundle with the following command:
```
deno run https://raw.githubusercontent.com/audrow/update-ros-pkg-versions/deploy/build/update-ros-pkg-versions.js --help
```

This is the error I get when trying to run the bundle:
```
error: Uncaught ReferenceError: Cannot access 'NodeTypeError' before initialization
class ERR_INVALID_ARG_TYPE extends NodeTypeError {
                                   ^
    at file:///Users/audrow/Documents/GitHub/update-ros-pkg-versions/out.js:4512:36
```

**Comment by ostera** [comment]
Here's a self-contained example that exhibits this same bug:

```javascript
// file: main.ts
import * as Fs from "https://deno.land/std@0.122.0/node/_fs/_fs_readFile.ts";                                                                   
Fs.readFileSync("./main.ts")
```

And here's how I run it:

```sh
; deno run -A main.ts
Check file:///private/tmp/fstest/main.ts
; echo $?
0
; deno compile main.ts
Check file:///private/tmp/fstest/main.ts
Bundle file:///private/tmp/fstest/main.ts
Compile file:///private/tmp/fstest/main.ts
Emit fstest
; ./fstest 
error: ReferenceError: Cannot access 'NodeTypeError' before initialization
    at file://$deno$/bundle.js:4418:36
```

**Comment by niedzielski** [comment]
I believe I have this same issue using `deno test`, not the bundler. It's something like:

1. I have a library index.ts which imports and re-exports many files like `export * from './utils/StrUtil.ts';`.
2. I have a test that imports the library index.ts.
3. When I execute `deno test`, I get errors like `error: ReferenceError: Cannot access 'I4XY' before initialization` and `error: TypeError: Cannot read properties of undefined (reading 'capitalize')`.
4. If I reorder the imports / exports from leaf dependency to root dependency, the tests work.

I believe I could isolate a minimal reproduction case but it will be time consuming for me. Please let me know if this is needed.

**Comment by niedzielski** [comment]
I think the issue is related to exporting both production code and assertSnapshot from the same mod.

**Comment by lucacasonato** [comment]
`deno bundle` is deprecated, this will not be fixed.
